Lut-free built-ins

cozdas opened the pull request:
- WIP Fixed-function for PQ<->Linear function. Doesn't have the GPU implementation yet and the rest may need tweaks.

cozdas committed:
- WIP Fixed-function for PQ<->Linear function. Doesn't have the GPU implementation yet and the rest may need tweaks.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

doug-walker reviewed (commented):
I need to review more thoroughly, but you're on the right track.

There's a lot more to adding a new FixedFunction though.  CTF support and Python support are required. Version checks must be updated to not allow the additions in config versions lower than 2.4.  Unit tests must be added to test all of that.

Some of the other PRs that might be a useful reference are #1404, #1814, and #896.

cozdas committed:
more progress on FIXED_FUNCTION_PQ_TO_LINEAR
- Python: Added docs
- Python: added the builtin function style
- Simplified the CPU code, it's still a prototype anyhow. Will optimize later.
- Style->OpData now takes the direction into account.
- Preliminary GPU implementation. This still clamps negative values.
- Added GPU tests.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- Fixed a range bug in the GPU test framework
- Added symmetry to the PQ curve both in CPU and GPU implementations
- Added version check for the FIXED_FUNCTION_PQ_TO_LINEAR transform.
- Adjusting the GPU test threshold as GPU and CPU (32f) are showing mismatch more than the distance each have to the ground truth.
- Adding fixed function OP tests
- Temporary change to the PQ curve CPU code for quickly switching between single and double precision floating point for error analysis.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas commented on src/OpenColorIO/transforms/builtins/Displays.cpp:
```diff
@@ -236,7 +192,8 @@ void RegisterAll(BuiltinTransformRegistryImpl & registry) noexcept
+            CreateFixedFunctionOp(ops, FixedFunctionOpData::PQ_TO_LINEAR, {});
         };
 
         registry.addBuiltin("CURVE - ST-2084_to_LINEAR",
```
Now we have the "CURVE - ST-2084_to_LINEAR" built-in transform and the PQ_TO_LINEAR built-in fixed function op. To me it feels like they are a bit redundant now. Are they?

cozdas replied:
I think my question was, if we didn't have the build-in transform, then ideally, we'd have only the fixed-function implementation which could be used the config files directly. Currently both the built-in transform and the fixed-function will be almost identical and we need to built-in transform just for backward compatibility.  Is my understanding correct?

doug-walker replied:
Yes, there will be both a Builtin Transform and FixedFunction that will be very similar in functionality but they are both required.

cozdas commented on tests/gpu/GPUUnitTest.cpp:
```diff
@@ -271,9 +271,9 @@ namespace
+            for (unsigned int i=0; i < numEntries; ++i, ++idx)
             {
-                tmp.m_inputValues[idx] = min + step * float(idx);
+                tmp.m_inputValues[idx] = min + step * float(i);
```
This was a bug: "step" was calculated using the remaining numEntries, but was multiplied by array index with the last element being (predefinedNumEntries-1). This was omitting the lower end values and over-shooting the specified max-value. This confused me a while.

cozdas committed:
- Adding CTF tests for PQ_TO_LINEAR FixedFunction.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

doug-walker reviewed (changes requested):
Please update tests/cpu/Config_tests.cpp, fixed_function_serialization, around line 4795 to validate Yaml serialization. Then add a check (there are other examples in that test) that an exception will be thrown on validation if the version is less than 2.4.

Please update tests/cpu/fileformats/FileFormatCTF_tests.cpp, ff_load_save_ctf, around line 3841 to validate CTF serialization.

I'm realizing that we have not been as rigorous on CTF versioning as we have on config versioning. Given that we are adding a lot for 2.4 we should bump the CTF version. This could be done in a follow-up.

cozdas commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpCPU.cpp:
```diff
@@ -1278,6 +1365,16 @@ ConstOpCPURcPtr GetFixedFunctionCPURenderer(ConstFixedFunctionOpDataRcPtr & func
+        }
+        case FixedFunctionOpData::LINEAR_TO_PQ:
+        {
+            // TODO: we may want to implement an SIMD renderer if scalar performance is low.
```
Agreed but please ignore all these cosmetic/style issues and temporary items for now. 

As I mentioned, the current state of the PR is for me to discover/learn all the parts that need to be updated for adding fixed-function built-ins and replacing the built-in transforms to use them. This is why I tried to keep this PR simple with only one sample function implemented (PQ), so that it'd be easier for both of us to iterate over this. Once we're happy with all the functional items in place, then I'll add the other built-in functions with proper formatting, and will do the code cleanup and beautifications. This is a WIP PR and trying to keep the formatting matching with the conventions (sometimes there seems to be multiple in a single file) is breaking the flow especially when my editor has different ideas about the conventions.

I think we need to come up with a clang format file and make it part of the repo, this way manual handing both for the code author and the reviewer will be unnecessary. I'm seeing that reformat requests are verry common in all PRs in the main repo too. Food for thought.

cozdas commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpCPU.cpp:
```diff
@@ -1178,7 +1194,78 @@ void Renderer_LUV_TO_XYZ::apply(const void * inImg, void * outImg, long numPixel
 
+namespace ST_2084
+{
+    using FLOAT = double; // Temp: used for fast float/double switching for precision evaluation.
```
This is a temp thing, please ignore for now.

cozdas commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpCPU.cpp:
I'm pretty sure that in this form the compiler will optimize it (unless we have very strict, non-standard floating point handling flags set for the compilers). I'll double check in the final form.

cozdas commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpGPU.cpp:
```diff
@@ -523,6 +523,59 @@ void Add_LUV_TO_XYZ(GpuShaderCreatorRcPtr & shaderCreator, GpuShaderText & ss)
+
+    ss.newLine() << ss.float3Decl("sign3") << " = sign(" << pxl << ".rgb);";
+    ss.newLine() << ss.float3Decl("x") << " = abs(" << pxl << ".rgb);";
+    ss.newLine() << "x = max(x, " << ss.float3Const(0.0) << ");";
```
Yup, leftover from the pre-symmetry implementation.  good catch.

cozdas commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpGPU.cpp:
```diff
@@ -523,6 +523,59 @@ void Add_LUV_TO_XYZ(GpuShaderCreatorRcPtr & shaderCreator, GpuShaderText & ss)
+    using namespace ST_2084;
+    const std::string pxl(shaderCreator->getPixelName());
+
+    // TODO: this still clamps negative inputs
```
leftover too, will be cleaned in the final cleanup pass.

cozdas commented on src/OpenColorIO/transforms/builtins/Displays.cpp:
```diff
@@ -22,50 +22,6 @@ namespace OCIO_NAMESPACE
 namespace DISPLAY
 {
 
```
We can always bring that back, in the final form of the PR based on the perf test results. I tried to keep the PR body short for easier review/iteration.

cozdas commented on tests/gpu/FixedFunctionOp_test.cpp:
```diff
@@ -505,3 +505,26 @@ OCIO_ADD_GPU_TEST(FixedFunction, style_XYZ_TO_LUV_inv)
+        OCIO::FixedFunctionTransform::Create(OCIO::FIXED_FUNCTION_PQ_TO_LINEAR);
+    func->setDirection(OCIO::TRANSFORM_DIR_FORWARD);
+
+    test.setTestWideRange(false);
```
Yes, it blows up, thus the narrow-range test. Maybe we can also utilize the ExpectedMinimalValue to keep the large errors under threshold, but extending the test-framework to accept custom ranges will be a better, more generic and controllable approach in my opinion too.

cozdas commented on tests/cpu/fileformats/FileFormatCTF_tests.cpp:
```diff
@@ -3838,6 +3838,8 @@ OCIO_ADD_TEST(FileFormatCTF, ff_load_save_ctf)
     ValidateFixedFunctionStyleNoParam(OCIO::FixedFunctionOpData::uvY_TO_XYZ         , __LINE__);
     ValidateFixedFunctionStyleNoParam(OCIO::FixedFunctionOpData::XYZ_TO_LUV         , __LINE__);
     ValidateFixedFunctionStyleNoParam(OCIO::FixedFunctionOpData::LUV_TO_XYZ         , __LINE__);
+    ValidateFixedFunctionStyleNoParam(OCIO::FixedFunctionOpData::PQ_TO_LINEAR       , __LINE__);
```
@doug-walker 
You requested CTF serialization implementation: "Please update tests/cpu/fileformats/FileFormatCTF_tests.cpp, ff_load_save_ctf, around line 3841 to validate CTF serialization."

This test was already extended to cover the new fixed-functions. What else do we want to test here?

cozdas committed:
- extending fixed_function_serialization test for PQ_TO_LINEAR fixedfunction
- minor fixes

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- GetFixedFunctionCPURenderer() now takes a parameter for fastpower function
- implemented SSE versions of the PQ curve in both directions with and without fastPower
- restored the old LUT-based PQ curve implementation that's used in the ST2084 Display builtin transform
- added custom range options to gpu unit tests
- changed the PQ curve test ranges to [-0.1, 1.1] in forward and [-0.1, 1001.1] in reverse directions

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- minor cleanup

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- OglApp::redisplay() function now can return the rendering time in nanoseconds, this is optional, so the existing clients don't need to change.
- HLG curve in CIE_XYZ_D65_to_REC2100_HLG_1000nit_Functor is pulled into its own function to be reused later on.
- HLG curve is no longer clipping at 0, it mirrors instead. clipping at 1 is currently untouched, may need to remove that too.
- fine-tuned the unit test thresholds for PQ curve so that they pass in all build/use situations.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- hopefully fixing the Windows builds that turn off SSE2 support but gcc and clang will keep on failing as I realized that those compilers don't have the _mm_pow_ps() SVML library implementation as msvc does. Will need to change the logic a bit.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- more attempts to make all builds happy

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- re-worked the logic a bit more so that non-windows systems will either use the fast-SSE or scalar implementation, while Windows will be able to use the precise-SSE implementation too if needed. Fingers crossed.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

doug-walker commented on src/OpenColorIO/transforms/builtins/Displays.cpp:
```diff
@@ -25,46 +29,87 @@ namespace DISPLAY
 namespace ST_2084
 {
 
```
The OCIO coding style does not indent for namespace.

doug-walker commented on src/OpenColorIO/transforms/builtins/Displays.cpp:
```diff
@@ -25,46 +29,87 @@ namespace DISPLAY
+                }
+                else
+                {
+                    // TODO: do we want remove clamp at 1 too per issue #1968 ?
```
Remove the clamp on both ends.

doug-walker commented on src/apps/ocioconvert/main.cpp:
```diff
@@ -450,10 +450,19 @@ int main(int argc, const char **argv)
+                std::cout << std::endl;
+                std::cout << "GPU processing took: "
+                    << 1.e-3f * renderDuration << " microseconds" << std::endl;
+            }
```
This would be better in ocioperf, but no strong objection, if it helps you.

cozdas replied:
Since ocioconvert is reporting the render duration in the verbose mode for CPU, I added it for GPU too to make it consistent. 

But I hit a problem with MacOS. Looks like GL_TIME_ELAPSED, is not implemented by MacOS even though it's part of  9-year old OpenGL 3.2 spec. Probably it's why this was not implemented in either of the command line tools in ocio. I had to roll this feature back :/

doug-walker commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpCPU.cpp:
Still needs to move.

cozdas replied:
Yeah this PR is still in the draft mode, doing the final cleanups now. I'll poke you when it's ready for another round.

doug-walker commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpCPU.cpp:
```diff
@@ -1278,6 +1514,44 @@ ConstOpCPURcPtr GetFixedFunctionCPURenderer(ConstFixedFunctionOpDataRcPtr & func
+        case FixedFunctionOpData::PQ_TO_LINEAR:
+        {
+#if OCIO_USE_SSE2
+            if(CPUInfo::instance().hasSSE2())
```
We assume all CPUs have at least SSE2, so this check is confusing. If people don't want that, they need to not set OCIO_USE_SSE2.

cozdas replied:
We have the same logic in Lut1DOpCPU.cpp line 283 and Lut3DOpCPU.cpp line 390. 
Do you want me to remove those run-time-checks too?

doug-walker replied:
We never had that until Mark Reid implemented all the AVX stuff. It kind of makes sense in that context where he has all the different versions. But there are a bunch of other places in the code that simply assume that if OCIO_USE_SSE2 that SSE2 may be used.  If the extra function call only happens every scan line, it's probably ok, but some people do use the scalar applyRGB to process images, so in that scenario it might add a significant penalty?

cozdas replied:
Normally it should be a single, inlined mask check but the way it's implemented currently doesn't seem to be inline-able. performance wise it should still be unnoticeable amongst all the other prep work we do until we reach here but these checks are not very useful indeed while the rest of the system assume SSE is available at run-time.  Though it would be a cleaner/consistent approach if we had the same logic for SSE and AVX and all the other instruction sets. 

I'm removing these checks then. Not touching the preexisting Lut code though.

doug-walker commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpCPU.cpp:
```diff
@@ -1278,6 +1514,44 @@ ConstOpCPURcPtr GetFixedFunctionCPURenderer(ConstFixedFunctionOpDataRcPtr & func
+                // on Windows we can use _mm_pow_ps() SVML "seuqential"
+                // intrinsic which is slower than our ssePower but precise. This
+                // will still be faster than scalar implementation.
+                return std::make_shared<Renderer_PQ_TO_LINEAR_SSE<false>>(func);
```
Is seuqential a word?

cozdas replied:
LOL no! good catch of my typo :)

doug-walker commented on tests/cpu/ops/fixedfunction/FixedFunctionOpCPU_tests.cpp:
```diff
@@ -557,3 +557,46 @@ OCIO_ADD_TEST(FixedFunctionOpCPU, XYZ_TO_LUV)
+        = std::make_shared<OCIO::FixedFunctionOpData>(OCIO::FixedFunctionOpData::LINEAR_TO_PQ);
+
+    img = outputFrame;
+    ApplyFixedFunction(&img[0], &inputFrame[0], NumPixels, dataFInv, 1e-5f, __LINE__);
```
It would be nice to test both the fastpow and non-fastpow implementations.

cozdas committed:
- Turns out mac OS does not implement the GL_TIME_ELAPSED query target which should be available since GL version 3.3 according to Khronos. Joy of cross-platform development :) Therefore removing the GPU render timing code from this PR. This is probably why GPU performance was not implemented in ocioperf already.

cozdas committed:
- Making gcc happy about the unused parameter. Probably it's a good idea to try to make all the compiler warning levels same/similar, some platforms being more picky about the same code than the others does not make much sense from the project's point of view, just frustration.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- One more step to make all compilers happy.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- Another pass to make clang happy.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- increasing the PQ_TO_LINEAR test threshold as ARM cpu seems to deviate more.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>

cozdas committed:
- removed the run-time checks for sse2 instruction set availability per code review discussion
- BT_2100 HGL clamp at 1 is removed.
- Fixed function tests can now ask for fast math on or off explicitly.
- Fixed function PQ curve tests are now performed both with and without fast math. With fast math on, the thresholds are increased to accommodate the accumulated error during roundtrip.

Signed-off-by: cuneyt.ozdas <cuneyt.ozdas@autodesk.com>
